compiler wrappers change implemented in #35912 & #35929 breaks builds of a few packages

### Steps to reproduce

The change introduced in https://github.com/spack/spack/pull/35912 PR breaks,

- openmpi%nag
- fgsl%intel

Also, the subsequent change introduced in https://github.com/spack/spack/pull/35929 PR breaks,

- likwid%intel

FYI, When I do not use commits from the above-mentioned PRs, the installation is a success.

This is how the compiler.yaml looks like, 

```
compilers::
#GCC
- compiler:
    spec: gcc@12.2.0
    paths:
      cc: /spack/opt/x86_64/gcc/12.2.0-gcc-coa32hm/bin/gcc
      cxx: /spack/opt/x86_64/gcc/12.2.0-gcc-coa32hm/bin/g++
      f77: /spack/opt/x86_64/gcc/12.2.0-gcc-coa32hm/bin/gfortran
      fc: /spack/opt/x86_64/gcc/12.2.0-gcc-coa32hm/bin/gfortran
    flags: {}
    operating_system: sles15
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
- compiler:
    spec: intel@2023.1.0
    paths:
      cc: /spack/opt/x86_64/intel-oneapi-compilers/2023.1.0-gcc-ncsrl25/compiler/2023.1.0/linux/bin/intel64/icc
      cxx: /spack/opt/x86_64/intel-oneapi-compilers/2023.1.0-gcc-ncsrl25/compiler/2023.1.0/linux/bin/intel64/icpc
      f77: /spack/opt/x86_64/intel-oneapi-compilers/2023.1.0-gcc-ncsrl25/compiler/2023.1.0/linux/bin/intel64/ifort
      fc: /spack/opt/x86_64/intel-oneapi-compilers/2023.1.0-gcc-ncsrl25/compiler/2023.1.0/linux/bin/intel64/ifort
    flags: {}
    operating_system: sles15
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
- compiler:
    spec: nag@7.1
    paths:
      cc: /spack/opt/x86_64/gcc/12.2.0-gcc-coa32hm/bin/gcc
      cxx: /spack/opt/x86_64/gcc/12.2.0-gcc-coa32hm/bin/g++
      f77: /spack/opt/x86_64/nag/7.1-gcc-gcpyzq7/bin/nagfor
      fc: /spack/opt/x86_64/nag/7.1-gcc-gcpyzq7/bin/nagfor
    flags:
       fflags: -mismatch
       fclags: -mismatch
    operating_system: sles15
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
```

I will be happy to provide further info if needed.

### Error message

```console
$> spack install openmpi%nag

==> Installing openmpi-4.1.5-pj344eofg6x6fk74wdac7yjvcghhbqzt
==> No binary for openmpi-4.1.5-pj344eofg6x6fk74wdac7yjvcghhbqzt found: installing from source
==> Using cached archive: /lrz/sys/spack/cache/_source-cache/archive/a6/a640986bc257389dd379886fdae6264c8cfa56bc98b71ce3ae3dfbd8ce61dbe3.tar.bz2
==> No patches needed for openmpi
==> openmpi: Executing phase: 'autoreconf'
==> openmpi: Executing phase: 'configure'
==> openmpi: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j24' 'V=1'

13 errors found in build log:

     23784    NAG Fortran Compiler Release 7.1(Hanzomon) Build 7101
  >> 23785    gcc: error: /dss/dsshome1/lrz/sys/spack/release/23.1.0/opt/haswell/openmpi/4.1.5-nag-pj344eo/lib: No such file or directory
  >> 23786    gcc: error: libmpi_usempi.so.40: No such file or directory
  >> 23787    gcc: error: unrecognized command line option '-Wl'; did you mean '-W'?
  >> 23788    gcc: error: unrecognized command line option '-Wl'; did you mean '-W'?
  >> 23789    gcc: error: unrecognized command line option '-Wl'; did you mean '-W'?
  >> 23790    gcc: error: unrecognized command line option '-Wl'; did you mean '-W'?
  >> 23791    gcc: error: unrecognized command line option '-Wl'; did you mean '-W'?
  >> 23792    gcc: error: unrecognized command line option '-Wl'; did you mean '-W'?
  >> 23793    gcc: error: unrecognized command line option '-soname'
  >> 23794    gcc: error: unrecognized command line option '-Wl'; did you mean '-W'?
  >> 23795    make[2]: *** [Makefile:1977: libmpi_usempi.la] Error 1
     23796    make[2]: Leaving directory '/gpfs/scratch/pr23ru/di39lit/di39lit/haswell/spack-stage-openmpi-4.1.5-pj344eofg6x6fk74wdac7yjvcghhbqzt/spack-src/ompi/mpi/fortran/use-mpi-tkr
              '
  >> 23797    make[1]: *** [Makefile:3566: all-recursive] Error 1
     23798    make[1]: Leaving directory '/gpfs/scratch/pr23ru/di39lit/di39lit/haswell/spack-stage-openmpi-4.1.5-pj344eofg6x6fk74wdac7yjvcghhbqzt/spack-src/ompi'
  >> 23799    make: *** [Makefile:1912: all-recursive] Error 1
```

```console
$> spack install likwid%intel

==> Installing likwid-5.1.1-xf637ed4ksq4oyzbpsibqev5l7tiysg6
==> No binary for likwid-5.1.1-xf637ed4ksq4oyzbpsibqev5l7tiysg6 found: installing from source
==> Using cached archive: /lrz/sys/spack/cache/_source-cache/archive/fa/faec7c62987967232f476a6ff0ee85af686fd24b5a360126896b7f435d1f943f.tar.gz
==> Ran patch() for likwid
==> likwid: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j24'

3 errors found in build log:
  >> 232    ld: cannot find /dss/dsshome1/lrz/sys/spack/release/23.1.0/opt/haswell/likwid/5.1.1-intel-xf637ed/lib: No such file or directory
  >> 233    make[1]: *** [Makefile:67: liblikwid-hwloc.so] Error 1
  >> 234    make: *** [Makefile:268: /gpfs/scratch/pr23ru/di39lit/di39lit/haswell/spack-stage-likwid-5.1.1-xf637ed4ksq4oyzbpsibqev5l7tiysg6/spack-src/ext/hwloc/liblikwid-hwloc.so] Erro
            r 2
```

```console
$> spack install fgsl%nag
==> Installing fgsl-1.5.0-vuz7g5za5v3c3bsmsfviq6kfzctyc5f2
==> No binary for fgsl-1.5.0-vuz7g5za5v3c3bsmsfviq6kfzctyc5f2 found: installing from source
==> Using cached archive: /lrz/sys/spack/cache/_source-cache/archive/50/5013b4e000e556daac8b3c83192adfe8f36ffdc91d1d4baf0b1cb3100260e664.tar.gz
==> No patches needed for fgsl
==> fgsl: Executing phase: 'autoreconf'
==> fgsl: Executing phase: 'configure'
==> fgsl: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j1' 'V=1'

11 errors found in build log:
            t/haswell/gsl/2.6-nag-cnrme7u/share/aclocal --include=/dss/dsshome1/lrz/sys/spack/release/23.1.0/opt/haswell/libtool/2.4.7-nag-et5pbfn/share/aclocal --include=/dss/dsshome1
            /lrz/sys/spack/release/23.1.0/opt/haswell/pkgconf/1.8.0-nag-akr6ygb/share/aclocal --force
     28     autoreconf: running: automake --add-missing --copy --force-missing
  >> 29     configure.ac:22: installing 'build-aux/ar-lib'
  >> 30     configure.ac:20: installing 'build-aux/compile'
  >> 31     configure.ac:8: installing 'build-aux/config.guess'
  >> 32     configure.ac:8: installing 'build-aux/config.sub'
  >> 33     configure.ac:6: installing 'build-aux/install-sh'
  >> 34     configure.ac:6: installing 'build-aux/missing'

     260    NAG Fortran Compiler Release 7.1(Hanzomon) Build 7101
  >> 261    gcc: error: libfgsl.so.1: No such file or directory
  >> 262    gcc: error: unrecognized command line option '-Wl'; did you mean '-W'?
  >> 263    gcc: error: unrecognized command line option '-soname'
  >> 264    gcc: error: unrecognized command line option '-Wl'; did you mean '-W'?
  >> 265    make: *** [Makefile:634: libfgsl.la] Error 1

```

### Information on your system

* **Spack:** 0.20.0.dev0 (8143c086a044d4afcd674161536da48a79a91674)
* **Python:** 3.6.15
* **Platform:** linux-sles15-skylake_avx512
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

--- a/lib/spack/env/cc
+++ b/lib/spack/env/cc
@@ -448,7 +448,9 @@ parse_Wl() {
         case "$1" in
             -rpath=*)
                 arg="${1#-rpath=}"
-                if system_dir "$arg"; then
+                if [ -z "$arg" ]; then
+                    shift; continue
+                elif system_dir "$arg"; then
                     append system_rpath_dirs_list "$arg"
                 else
                     append rpath_dirs_list "$arg"
@@ -456,7 +458,9 @@ parse_Wl() {
                 ;;
             --rpath=*)
                 arg="${1#--rpath=}"
-                if system_dir "$arg"; then
+                if [ -z "$arg" ]; then
+                    shift; continue
+                elif system_dir "$arg"; then
                     append system_rpath_dirs_list "$arg"
                 else
                     append rpath_dirs_list "$arg"

--- a/lib/spack/env/cc
+++ b/lib/spack/env/cc
@@ -433,6 +433,8 @@ wl_expect_rpath=no
 # Same, but for -Xlinker -rpath -Xlinker /path
 xlinker_expect_rpath=no
 
+wl_prefix=-Wl,
+
 parse_Wl() {
     while [ $# -ne 0 ]; do
     if [ "$wl_expect_rpath" = yes ]; then
@@ -470,12 +472,19 @@ parse_Wl() {
             "$dtags_to_strip")
                 ;;
             *)
-                append other_args_list "-Wl,$1"
+                append other_args_list "$wl_prefix$1"
                 ;;
         esac
     fi
     shift
     done
+}
+
+parse_Wl_NAG() {
+    # For -Wl,-Wl,,x,,y,,z; we receive $@ = [x, '', y, '', z]
+    # so dropping empty arguments with unset IFS gives 'parse_Wl x y z'
+    unset IFS
+    parse_Wl $@
 }
 
 
@@ -577,10 +586,17 @@ while [ $# -ne 0 ]; do
             append other_args_list "-l$arg"
             ;;
         -Wl,-Wl,,*)
-            # We never dealt with the NAG compiler -Wl,-Wl,,-rpath,,/path construct
-            append other_args_list "$1"
+            # NAG Fortran compiler linker flags are of the form -Wl,-Wl,,x,,y,,z
+            # Here we split on comma, and let parse_Wl_NAG drop empty arguments.
+            # This is still buggy, since '-Wl,-Wl,,-rpath,,/path with spaces' fails,
+            # but it's better than nothing.
+            wl_prefix=-Wl,-Wl,,
+            IFS=,
+            parse_Wl_NAG ${1#-Wl,-Wl,,}
+            unset IFS
             ;;
         -Wl,*)
+            wl_prefix=-Wl,
             IFS=,
             parse_Wl ${1#-Wl,}
             unset IFS

--- a/lib/spack/env/cc
+++ b/lib/spack/env/cc
@@ -577,8 +577,7 @@ while [ $# -ne 0 ]; do
             append other_args_list "-l$arg"
             ;;
         -Wl,-Wl,,*)
-            # We never dealt with NAG compiler -Wl,-Wl,,-rpath,,/path construct; they are
-            # ignored here, and instead we hope that the linker gets intercepted.
+            # We never dealt with the NAG compiler -Wl,-Wl,,-rpath,,/path construct
             append other_args_list "$1"
             ;;
         -Wl,*)

--- a/lib/spack/env/cc
+++ b/lib/spack/env/cc
@@ -434,8 +434,6 @@ wl_expect_rpath=no
 xlinker_expect_rpath=no
 
 parse_Wl() {
-    # drop -Wl
-    shift
     while [ $# -ne 0 ]; do
     if [ "$wl_expect_rpath" = yes ]; then
         if system_dir "$1"; then
@@ -580,7 +578,7 @@ while [ $# -ne 0 ]; do
             ;;
         -Wl,*)
             IFS=,
-            parse_Wl $1
+            parse_Wl ${1#-Wl,}
             unset IFS
             ;;
         -Xlinker)

--- a/lib/spack/env/cc
+++ b/lib/spack/env/cc
@@ -471,6 +471,11 @@ parse_Wl() {
                 ;;
             "$dtags_to_strip")
                 ;;
+            -Wl)
+                # Nested -Wl,-Wl means we're in NAG compiler territory, we don't support
+                # it.
+                return 1
+                ;;
             *)
                 append other_args_list "$wl_prefix$1"
                 ;;
@@ -478,13 +483,6 @@ parse_Wl() {
     fi
     shift
     done
-}
-
-parse_Wl_NAG() {
-    # For -Wl,-Wl,,x,,y,,z; we receive $@ = [x, '', y, '', z]
-    # so dropping empty arguments with unset IFS gives 'parse_Wl x y z'
-    unset IFS
-    parse_Wl $@
 }
 
 
@@ -585,20 +583,12 @@ while [ $# -ne 0 ]; do
             if [ -z "$arg" ]; then shift; arg="$1"; fi
             append other_args_list "-l$arg"
             ;;
-        -Wl,-Wl,,*)
-            # NAG Fortran compiler linker flags are of the form -Wl,-Wl,,x,,y,,z
-            # Here we split on comma, and let parse_Wl_NAG drop empty arguments.
-            # This is still buggy, since '-Wl,-Wl,,-rpath,,/path with spaces' fails,
-            # but it's better than nothing.
-            wl_prefix=-Wl,-Wl,,
-            IFS=,
-            parse_Wl_NAG ${1#-Wl,-Wl,,}
-            unset IFS
-            ;;
         -Wl,*)
             wl_prefix=-Wl,
             IFS=,
-            parse_Wl ${1#-Wl,}
+            if ! parse_Wl ${1#-Wl,}; then
+                append other_args_list "$1"
+            fi
             unset IFS
             ;;
         -Xlinker)

--- a/lib/spack/env/cc
+++ b/lib/spack/env/cc
@@ -433,8 +433,6 @@ wl_expect_rpath=no
 # Same, but for -Xlinker -rpath -Xlinker /path
 xlinker_expect_rpath=no
 
-wl_prefix=-Wl,
-
 parse_Wl() {
     while [ $# -ne 0 ]; do
     if [ "$wl_expect_rpath" = yes ]; then
@@ -477,7 +475,7 @@ parse_Wl() {
                 return 1
                 ;;
             *)
-                append other_args_list "$wl_prefix$1"
+                append other_args_list "-Wl,$1"
                 ;;
         esac
     fi
@@ -584,7 +582,6 @@ while [ $# -ne 0 ]; do
             append other_args_list "-l$arg"
             ;;
         -Wl,*)
-            wl_prefix=-Wl,
             IFS=,
             if ! parse_Wl ${1#-Wl,}; then
                 append other_args_list "$1"

--- a/lib/spack/env/cc
+++ b/lib/spack/env/cc
@@ -576,6 +576,11 @@ while [ $# -ne 0 ]; do
             if [ -z "$arg" ]; then shift; arg="$1"; fi
             append other_args_list "-l$arg"
             ;;
+        -Wl,-Wl,,*)
+            # We never dealt with NAG compiler -Wl,-Wl,,-rpath,,/path construct; they are
+            # ignored here, and instead we hope that the linker gets intercepted.
+            append other_args_list "$1"
+            ;;
         -Wl,*)
             IFS=,
             parse_Wl ${1#-Wl,}

--- a/lib/spack/spack/test/cc.py
+++ b/lib/spack/spack/test/cc.py
@@ -352,6 +352,7 @@ def test_Wl_parsing(wrapper_environment):
     )
 
 
+@pytest.mark.regression("37179")
 def test_Wl_parsing_with_missing_value(wrapper_environment):
     check_args(
         cc,
@@ -360,6 +361,7 @@ def test_Wl_parsing_with_missing_value(wrapper_environment):
     )
 
 
+@pytest.mark.regression("37179")
 def test_Wl_parsing_NAG_is_ignored(wrapper_environment):
     check_args(
         fc,

--- a/lib/spack/spack/test/cc.py
+++ b/lib/spack/spack/test/cc.py
@@ -351,13 +351,12 @@ def test_Wl_parsing(wrapper_environment):
         + ["-Wl,--disable-new-dtags", "-Wl,-rpath,/a", "-Wl,-rpath,/b", "-Wl,-rpath,/c"],
     )
 
+
 def test_Wl_parsing_with_missing_value(wrapper_environment):
     check_args(
         cc,
         ["-Wl,-rpath=/a,-rpath=", "-Wl,--rpath="],
-        [real_cc]
-        + target_args
-        + ["-Wl,--disable-new-dtags", "-Wl,-rpath,/a"],
+        [real_cc] + target_args + ["-Wl,--disable-new-dtags", "-Wl,-rpath,/a"],
     )
 
 

--- a/lib/spack/spack/test/cc.py
+++ b/lib/spack/spack/test/cc.py
@@ -360,6 +360,14 @@ def test_Wl_parsing_with_missing_value(wrapper_environment):
     )
 
 
+def test_Wl_parsing_NAG_is_ignored(wrapper_environment):
+    check_args(
+        fc,
+        ["-Wl,-Wl,,x,,y,,z"],
+        [real_cc] + target_args + ["-Wl,--disable-new-dtags", "-Wl,-Wl,,x,,y,,z"],
+    )
+
+
 def test_Xlinker_parsing(wrapper_environment):
     # -Xlinker <x> ... -Xlinker <y> may have compiler flags inbetween, like -O3 in this
     # example. Also check that a trailing -Xlinker (which is a compiler error) is not

--- a/lib/spack/spack/test/cc.py
+++ b/lib/spack/spack/test/cc.py
@@ -351,6 +351,15 @@ def test_Wl_parsing(wrapper_environment):
         + ["-Wl,--disable-new-dtags", "-Wl,-rpath,/a", "-Wl,-rpath,/b", "-Wl,-rpath,/c"],
     )
 
+def test_Wl_parsing_with_missing_value(wrapper_environment):
+    check_args(
+        cc,
+        ["-Wl,-rpath=/a,-rpath=", "-Wl,--rpath="],
+        [real_cc]
+        + target_args
+        + ["-Wl,--disable-new-dtags", "-Wl,-rpath,/a"],
+    )
+
 
 def test_Xlinker_parsing(wrapper_environment):
     # -Xlinker <x> ... -Xlinker <y> may have compiler flags inbetween, like -O3 in this

